clang/lib/Driver/ToolChains/Darwin.cpp

```diff
@@ -1018,13 +1018,19 @@ bool Darwin::hasBlocksRuntime() const {
   }
 }
 
-void Darwin::AddCudaIncludeArgs(const ArgList &DriverArgs,
-                                ArgStringList &CC1Args) const {
+void MachO::AddCudaIncludeArgs(const ArgList &DriverArgs,
+                               ArgStringList &CC1Args) const {
+  if (!isTargetAppleBased())
+    return ToolChain::AddCudaIncludeArgs(DriverArgs, CC1Args);
```

Comment:
Having to check for `isTargetAppleBased` all over the place makes me wonder whether the inheritance hierarchy has the right shape.

Under what concrete triples are these code paths reached?

Reply:
From `Driver::getToolChain` you get it for these
arch-vendor-win32-macho
arch-vendor-windows-macho
arch-vendor-none-macho

The only one I really want to change is arch-apple-none-macho, I don't want to disturb any of the others. But that's a good point, it would probably be cleaner to make an AppleMachO toolchain in between the existing MachO and Darwin ones

Reply:
> But that's a good point, it would probably be cleaner to make an AppleMachO toolchain in between the existing MachO and Darwin ones

This does seem cleaner to me, but I can see how someone else might prefer the opposite -- feel free to leave the `isTargetAppleBased` checks in if someone else pushes back on an `AppleMachO` toolchain.

Reply:
@jroelofs what do you think? Personally I prefer an `AppleMachO` toolchain instead of sprinkling `isTargetAppleBased` in a bunch of places.

Reply:
`AppleMachO` seems like a good idea to me.